File: java/src/main/java/org/mydotey/quantileestimator/decorator/ConcurrencyDecorator.java
package org.mydotey.quantileestimator.decorator;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.locks.ReentrantLock;

import org.mydotey.quantileestimator.QuantileEstimator;

/**
 * @author koqizhao
 *
 * Apr 2, 2018
 */
public class ConcurrencyDecorator<T> implements QuantileEstimator<T> {

    private QuantileEstimator<T> _quantileEstimator;

    private ConcurrentLinkedQueue<T> _cache;
    private ReentrantLock _lock;

    public ConcurrencyDecorator(QuantileEstimator<T> quantileEstimator) {
        Objects.requireNonNull(quantileEstimator, "quantileEstimator is null");

        _quantileEstimator = quantileEstimator;
        _cache = new ConcurrentLinkedQueue<>();
        _lock = new ReentrantLock();
    }

    @Override
    public void add(T value) {
        _cache.offer(value);

        if (_lock.tryLock()) {
            try {
                batchAdd();
            } finally {
                _lock.unlock();
            }
        }
    }

    @Override
    public Map<Double, T> get(List<Double> quantiles) {
        try {
            _lock.lock();

            batchAdd();
            return _quantileEstimator.get(quantiles);
        } finally {
            _lock.unlock();
        }
    }

    protected void batchAdd() {
        for (T cached = _cache.poll(); cached != null; cached = _cache.poll())
            _quantileEstimator.add(cached);
    }

}

File: dotnet/src/MyDotey.QuantileEstimator/Concurrent/ConcurrencyDecorator.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Threading;

/**
 * @author koqizhao
 *
 * Apr 2, 2018
 */
namespace MyDotey.Quantile.Concurrent
{
    public class ConcurrencyDecorator<T> : IQuantileEstimator<T>
    {
        private IQuantileEstimator<T> _quantileEstimator;

        private ConcurrentQueue<T> _cache;
        private object _lock;

        public ConcurrencyDecorator(IQuantileEstimator<T> quantileEstimator)
        {
            if (quantileEstimator == null)
                throw new ArgumentNullException("quantileEstimator is null");

            _quantileEstimator = quantileEstimator;
            _cache = new ConcurrentQueue<T>();
            _lock = new object();
        }

        public virtual void Add(T value)
        {
            _cache.Enqueue(value);

            if (Monitor.TryEnter(_lock))
            {
                try
                {
                    BatchAdd();
                }
                finally
                {
                    Monitor.Exit(_lock);
                }
            }
        }

        public virtual Dictionary<Double, T> Get(List<Double> quantiles)
        {
            lock (_lock)
            {
                BatchAdd();
                return _quantileEstimator.Get(quantiles);
            }
        }

        protected virtual void BatchAdd()
        {
            for (bool success = _cache.TryDequeue(out T cached); success; success = _cache.TryDequeue(out cached))
                _quantileEstimator.Add(cached);
        }
    }
}